_published/0017-azuredevops-repos-and-not-projects.md

```diff
@@ -0,0 +1,259 @@
+   ```
+
+   <https://docs.microsoft.com/en-us/rest/api/azure/devops/git/refs/list#refs-heads>
+
+   This is used to get the following information:
+
+   - Branch name (but make sure to trim away the `refs/heads/` prefix)
+
+## Compatibility
+
+### Renaming projects
+
+Due to the Wharf project name change we will have to add backward
+compatibility to find these projects when importing with the new version
+and automatically rename them. Something like:
+
+1. Try find Wharf project according rules defined in
+   [#Wharf project naming](#wharf-project-naming).
+
+   If there are more than 1 code repository, try with
+   `group="{azure-org}" && name="{azure-project}"`, as that was the old behavior
+   from wharf-provider-azuredevops v1.
+
+2. Update Wharf project with fresh data, including the new group name and
+   project name.
+
+We need to get through the ["group name cannot be changed"](https://github.com/iver-wharf/wharf-api/blob/74a4718fa830413a7e03cc5efed4d56f08f0398d/project.go#L365-L375)
+validation criteria. As the validation is arbitrary it will be removed.
+This would also resolve the inconsistency that the validation is not checked in
+the `POST /project` endpoint.
+
+### Build pipelines will fail
+
+Builds that solely rely on the [`REPO_GROUP`](https://iver-wharf.github.io/#/usage-wharfyml/variables/built-in-variables?id=repo_group)
+or [`REPO_NAME`](https://iver-wharf.github.io/#/usage-wharfyml/variables/built-in-variables?id=repo_name)
+variables, either explicitly through variable substitution or implicitly such
+as via the [`docker.group` or `docker.name`](https://iver-wharf.github.io/#/usage-wharfyml/step-types/docker)
+(undocumented) defaults that use the same variables behind the scenes, will be
+fine. Perhaps some Docker images are pushed to a new address, but as long as
+that new registry address is writable and wherever that address is referenced
+uses those above built-in variables then will work fine.
+
+The issue is where users have hardcoded these URLs into their continuous
+deployments. The Docker images will be pushed to a new location, but the old
+location is still what's referenced, causing their builds to break.
+
+This information needs to go out, preferably before this is released, to all
+known users of wharf-provider-azuredevops telling them to start using
+`${REPO_GROUP}` and `${REPO_NAME}` instead.
+
+For the unknown users, we should try and word the CHANGELOG.md as clear as
+possible declaring these impacts. We should also write a document on
+<https://iver-wharf.github.io/> about upgrading wharf-provider-azuredevops from
+v1 to v2, where we explain these backward incompatibilities as well.
+
+## Alternative solutions
+
+For the sake of resolving
+<https://github.com/iver-wharf/wharf-provider-azuredevops/issues/24> we could
+settle with still importing 1 Wharf project per Azure DevOps project, but
+this is so tightly coupled so I [@jilleJr] think it's best to tackle this
```

Comment:
nit: Form here as you are the author of the RFC and this is a techincal doc opinions should be reformulated as fact. 
I think debates should go into the `Unresolved questions`, the RFC should assume this debate has been done. The alternatives considered has always seem to be more of a inventory check to me. As in if we come back to this later and ask why didnt we consider x? Then we get that answer here in the form of a benefits drawback discussion or alternativly we get the answer "we didnt consider x at all".

Reply:
I like this proposed approach. Personally I've considered the "Alternatives" section as pointing out the obvious alternatives (to begin with) so they can be dismissed during the review period. Were the author answers potential "why didn't you just do X?" questions.

But if I understand you correctly, we could also move the "unresolved questions" over to "alternatives" once they're answered/discussed? Because I like that.

Reply:
I think it would depend on the nature of the unresolved question but in principal yes. If the unresolved question is expected to be classified as an alternative solution in advance- i think i would expect a discusssion call/brainstorming meeting. Listing it as a discussion for a text forum such as a PR seems inefficient. Meetings are great for brainstorming. Once the upsides and downsides are listed as facts it could then go straight into the alternatives considered, or in the  alterntive case replace the old core idea as the target of the RFC.

Reply:
Yea cool, so I think we have the same mindset here. Correct me if I misunderstood, but here's my image:

- "Alternatives": document some resolved questions and other obvious misdirections the author intentionally didn't take
- "Unresolved questions": things to talk about in meeting or discuss in PRs, preferably verbally for speed and brainstorming

Reply:
I think the relevance of alternatives increase in the scenario where there are multiple good mutually exclusive options to move forward with. The discussion should be included for the sake of reference.